### ISSUE TITLE
improve styling of <details>/<summary> elements

### ISSUE THREAD

**Issue by dbaron** [w3c/csswg-drafts#9879]
I'd like to discuss improvements to the styling of the `<details>` and `<summary>` elements.

At our face-to-face meeting in July 2023, I presented [some thoughts](https://github.com/dbaron/details-styling) on improvements, and we had a discussion (see [minutes](https://lists.w3.org/Archives/Public/www-style/2023Sep/0019.html)).

I'm hoping to have some updates to present during our face-to-face in February, though they're not ready yet.

**Comment by dbaron** [comment]
Based on the investigations we've done so far into styling of the `<details>` and `<summary>` elements, I can think of 6 separate technical pieces that are relevant to this issue (items 4 through 6 are relevant to animations of opening/closing the elements), in somewhat different states:

1. Allowing general use of `display` on both `<details>` and `<summary>`
   * covered by whatwg/html#9830; technically an HTML issue rather than a CSS one
   * [prototyped in Chromium](https://chromestatus.com/feature/5112013093339136) behind "Experimental Web platform features"
   * **Proposed resolution**: express support for changing HTML to allow `details` and `summary` to be more styleable
2. Specifying the styles that create the disclosure triangle.
   * as discussed in https://github.com/dbaron/details-styling these differ between Chromium & Gecko on one side, and WebKit on the other side
   * having the underlying styles be consistent between implementations would allow authors to style more easily, and may also help with the previous point since WebKit's restrictions on `summary { display: list-item }` may be motivated by people writing styles to the Gecko/Chromium model
   * **Proposed resolution**: Adopt the Gecko/Chromium styling model in which the `summary` is `display: list-item` and the disclosure triangle is its marker.
3. Add a pseudo-element for styling the anonymous container for all of the non-`summary` contents of a `details` element
   * I have currently prototyped this as `::details-content`, but I'm open to a different name.
   * [prototyped in Chromium](https://chromestatus.com/feature/5112013093339136) behind "Experimental Web platform features"
   * **Discussion topic**: One thing that is awkward about using it is that it defaults to `display: contents`, which means that using it for anything requires changing it to `display: block`
   * **Proposed resolution**: Adopt the `::details-content` pseudo-element.
4. The `transition-behavior` property
   * [part of css-transitions-2](https://drafts.csswg.org/css-transitions-2/#transition-behavior-property) (per [resolution](https://github.com/w3c/csswg-drafts/issues/8857#issuecomment-1581650809) and [resolution](https://github.com/w3c/csswg-drafts/issues/8857#issuecomment-1646035541))
   * [shipped in Chromium](https://chromestatus.com/feature/5071230636392448)
   * part of the `@starting-style` and `transition-behavior` focus area in [Interop 2024](https://wpt.fyi/interop-2024)
5. The `@starting-style` rule
   * [part of css-transitions-2](https://drafts.csswg.org/css-transitions-2/#defining-before-change-style) (per [resolution](https://github.com/w3c/csswg-drafts/issues/8174#issuecomment-1515026543) and [resolution](https://github.com/w3c/csswg-drafts/issues/8174#issuecomment-1561484749))
   * [shipped in Chromium](https://chromestatus.com/feature/4515377717968896)
   * part of the `@starting-style` and `transition-behavior` focus area in [Interop 2024](https://wpt.fyi/interop-2024)
6. The ability to transition to/from auto heights.
   * This is covered in a separate issue, see current proposal in https://github.com/w3c/csswg-drafts/issues/626#issuecomment-1800254442
   * We have a [resolution](https://github.com/w3c/csswg-drafts/issues/626#issuecomment-1864906213) to add this to [css-values-5](https://drafts.csswg.org/css-values-5/), but the spec edits have not happened yet.
   * I'm [working on prototyping this in Chromium](https://chromestatus.com/feature/5196713071738880)

There was another piece that I had prototyped (closely related to item 3 above), but removed from the prototype because it turned out to not be useful.  This was the `::details-summary` pseudo-element which styled the container for the `<summary>` element.  Similar to `::details-content` it was a little tricky to use because it defaulted to `display: contents`, and it didn't seem to add any value, only confusion about when to use it and when not to.  (It's just a wrapper for a single element, the first summary child.  It's potentially useful in the case where there is no summary child, but I think that's mostly an error case anyway.)

**Comment by dbaron** [comment]
At this point the discussion in #9951 and the edits for #10083 have resolved on many of the issues that were blocking resolution here.  The one point that I don't think we resolved on (unless I missed it) was the *name* of the pseudo-element.  I'd like to resolve to use `::details-content` as the name for the pseudo-element for the part of the `<details>` that opens and closes.  I'm adding `Agenda+` for that, although it's possible we could also do it as an asynchronous resolution given that it's already been discussed a bit.

**Comment by astearns** [comment]
The CSSWG will automatically accept this resolution one week from now if no objections are raised here. Anyone can add an emoji to this comment to express support. If you do not support this resolution, please add a new comment.

Proposed Resolution: Use `::details-content` as the name for the pseudo-element for the part of `<details>` that opens and closes

**Comment by kbabbitt** [comment]
No objection to the proposal. The previous resolution called for a joint meeting with WHATWG; for anyone like me who was looking for the outcome of that discussion, it's in the minutes here: https://github.com/whatwg/html/issues/10200#issuecomment-1998676547

> RESOLUTION: We will use pseudo-elements for exposing styling of pieces of built-in controls. We would like these pseudo-elements to be as close to ::part() as possible (for example, in terms of what selectors work and how inheritance works) and intend to further explore the details of how to do this. (9 votes in favor, none opposed)

**Comment by yisibl** [comment]
I don't quite understand why there are only a [limited number of CSS properties](https://drafts.csswg.org/css-lists-3/#marker-properties) that can be used in `::marker`, which severely restricts details from being a truly customizable styled component. 

A very common use case is when arrows need to be animated using the `rotate` property.


- Ant design: https://ant-design.antgroup.com/components/collapse#components-collapse-demo-accordion

https://github.com/w3c/csswg-drafts/assets/2784308/add5858a-25b7-40cb-b2cb-9d3111920c58

- Bootstrap: https://getbootstrap.com/docs/5.0/components/accordion/#example

**Comment by yisibl** [comment]
Here is a collection of requests to support more properties in `::marker`.

- https://github.com/w3c/csswg-drafts/issues/8892
- https://github.com/w3c/csswg-drafts/issues/6203
- https://github.com/w3c/csswg-drafts/issues/5265
- https://github.com/w3c/csswg-drafts/issues/5250

**Comment by dbaron** [comment]
Yeah, I agree that we should probably make marker styling more flexible.  I haven't yet dug in to propose an exact set of changes...

**Comment by astearns** [comment]
RESOLVED: Use `::details-content` as the name for the pseudo-element for the part of `<details>` that opens and closes

**Comment by annevk** [comment]
Is it problematic that it's `display: contents` but `details-content`? I guess there's already "content" usage elsewhere that will be similarly hard to remember.

**Comment by bkardell** [comment]
> Is it problematic that it's display: contents but details-content? I guess there's already "content" usage elsewhere that will be similarly hard to remember.

Imo, no.
